django/template/base.py

```diff
@@ -154,23 +162,30 @@ class InvalidTemplateLibrary(Exception):
 
 
 class Origin(object):
-    def __init__(self, name):
+    def __init__(self, name, template_name=None, loader=None, dirs=None):
         self.name = name
-
-    def reload(self):
-        raise NotImplementedError('subclasses of Origin must provide a reload() method')
+        self.template_name = template_name
+        self.loader = loader
+        self.dirs = dirs
 
     def __str__(self):
         return self.name
 
+    def __eq__(self, other):
+        if not isinstance(other, Origin):
+            return False
 
-class StringOrigin(Origin):
-    def __init__(self, source):
-        super(StringOrigin, self).__init__(UNKNOWN_SOURCE)
-        self.source = source
+        return (
+            self.name == other.name and
+            self.loader == other.loader
+        )
 
-    def reload(self):
-        return self.source
+    @property
+    def loader_name(self):
+        if self.loader:
+            return '%s.%s' % (
+                self.loader.__module__, self.loader.__class__.__name__,
+            )
```

Comment:
If a non-Django engine was to provide postmortem information, we'd need to provide a base ``Origin`` class. The only required attributes are ``name`` and ``loader_name``, where ``name`` is the template identifier and ``loader_name`` is a string identifying the method used to loader the template.

Jinja2 doesn't look like it exposes that information, though. So I just kept things in ``django.template.base`` for now.